lib/rpc/src/tcp_client.rs

```diff
@@ -0,0 +1,463 @@
+        if data.len() > 0 {
+            //unsafe { encode(&data, &mut req_data) }.unwrap();
+            req_data.extend(data);
+        }
+
+        // Send request
+        self.msg_send(req_data).unwrap();
+
+        // Receive response header
+        let mut res_data = self.msg_recv(core::mem::size_of::<RPCHeader>()).unwrap();
+        let (res_hdr, extra) = unsafe { decode::<RPCHeader>(&mut res_data) }.unwrap();
+        assert_eq!(extra.len(), 0);
+
+        // Read the rest of the data
+        let mut payload_data = Vec::new();
+        if res_hdr.msg_len > 0 {
+            payload_data = self.msg_recv(res_hdr.msg_len as usize).unwrap();
+        }
+
+        // Check request & client IDs, and also length of received data
+        if ((res_hdr.client_id != self.client_id) && rpc_id != RPCType::Registration)
+            || res_hdr.req_id != self.req_id
+        {
+            warn!(
+                "Mismatched client id ({}, {}) or request id ({}, {})",
+                res_hdr.client_id, self.client_id, res_hdr.req_id, self.req_id
+            );
+            return Err(RPCError::MalformedResponse);
+        }
+
+        // Increment request id
+        self.req_id += 1;
+
+        // If registration, update id
+        if rpc_id == RPCType::Registration {
+            self.client_id = res_hdr.client_id;
+            debug!("Set client ID to: {}", self.client_id);
+            return Ok(Vec::new());
+        }
+        Ok(payload_data)
+    }
+
+    /// send data to a remote node
+    fn msg_send(&mut self, data: Vec<u8>) -> Result<(), RPCError> {
+        let mut data_sent = 0;
+        loop {
+            match self.iface.poll(&mut self.sockets, Instant::from_millis(0)) {
+                Ok(_) => {}
+                Err(e) => {
+                    warn!("poll error: {}", e);
+                }
+            }
+
+            if data_sent == data.len() {
+                return Ok(());
+            } else {
+                let mut socket = self.sockets.get::<TcpSocket>(self.server_handle.unwrap());
+                if socket.can_send() && socket.send_capacity() > 0 && data_sent < data.len() {
+                    let end_index = data_sent + core::cmp::min(data.len(), socket.send_capacity());
+                    debug!("msg_send [{:?}-{:?}]", data_sent, end_index);
+                    if let Ok(bytes_sent) = socket.send_slice(&data[data_sent..end_index]) {
+                        trace!(
+                            "Client sent: [{:?}-{:?}] {:?}/{:?} bytes",
+                            data_sent,
+                            end_index,
+                            end_index,
+                            data.len()
+                        );
+                        data_sent = data_sent + bytes_sent;
+                    } else {
+                        debug!("send_slice failed... trying again?");
+                    }
+                }
+            }
+        }
+    }
```

Comment:
here as well, `data` is deallocated at the end of this scope but might be better if we can return the buffer so a client can re-use for the next send?